nexus/examples/config.toml

```diff
@@ -47,3 +47,8 @@ mode = "stderr-terminal"
 #mode = "file"
 #path = "logs/server.log"
 #if_exists = "append"
+
+[dtrace]
+# If `true`, Nexus will abort on startup if the embedded DTrace probes cannot
+# be registered.
+required = true
```

Comment:
I believe the README recommends running Neuxs in this section: https://github.com/oxidecomputer/omicron#build-and-run

Would this no longer work on Linux systems with this updated config?

FWIW, if you wanted to add this to the [SMF config](https://github.com/oxidecomputer/omicron/blob/main/smf/nexus/config.toml), which is used exclusively on illumos, IMO that would be fine.

Reply:
This will still work on Linux. The `usdt` crate selects an empty implementation on platforms without DTrace, and the "registration" function always returns `Ok`.

But I'm curious if there's a better way to specify this. I think it's important to avoid accidentally disabling the probes or not noticing that registration fails. The config file seemed the least error-prone, but a CLI flag might be OK too, especially if the flag is `--no-require-dtrace` so that the default is to abort if registration fails.

Reply:
Nice catch on the SMF config, will do. I didn't realize that was an additional file! I thought the existing `nexus/examples/config.toml` was copied into the SMF tarball.

Reply:
Ah, so if this can work on Linux, Mac, and illumos, why bother leaving it as a config? Why let someone ignore this flavor of error?

Reply:
I think there are situations where one doesn't care. The most likely situation in which registration fails is that the DTrace kernel module is out of memory. In this degraded state, we might care more about the program (say, sled agent) running at all, and not want to block the system being partly functional just because we don't have DTrace probes. But in the normal case, we'd want to avoid silently ignoring such a failure. Logging an error message seems the least we can do, but going further by aborting the program seems heavy-handed unless we've explicitly requested that behavior.

Does that all seem reasonable? Do you see a better approach here?

Reply:
(warning; opinion)

That seems like a hard line to draw though, and since this is a configuration option, when would we practically deploy that? Like, when should we ship a config that says "dtrace probe failures are OK"?

I would think for the "normal" case that if dtrace probes cannot be registered, we'd wanna know! I know you mentioned that this would appear in the logs, but that seems unlikely to be noticed unless someone is explicitly looking for it.

Arguably, a kernel module running out of memory could fail a *LOT* of operations in the sled agent, and I (personally) would rather have explicit+loud errors than silent ones.

Reply:
> That seems like a hard line to draw though, and since this is a configuration option, when would we practically deploy that? Like, when should we ship a config that says "dtrace probe failures are OK"?

That's a good point. I was trying to think of how we'd switch this option. A config parameter seems less flexible than a CLI flag, but neither is really "easy". We'd need to have access that permits restarting the process with a modified config/argv, which seems...problematic. So I agree.

> I (personally) would rather have explicit+loud errors than silent ones.

Definitely! I just don't know what we _do_ with that error. Logging the message is the minimal action, but requires someone looking as you point out. And it's true that the DTrace kernel module running low on memory might imply a lot of other operations are fragile or degraded. But it does seem like the probes should be first on the chopping block. And aborting the entire program seems a less graceful failure mode than provided degraded service.

One possibility is generating an alert or FMA event when registration fails, but _not_ aborting the program. @davepacheco What do you think here? Ultimately we need to make a choice to abort or not, but I do want to avoid providing a useless knob in the form of a configuration bit that we can't twiddle. Maybe aborting is good at this early stage, with the intent to change this to an alert or FMA event in the future.

Reply:
The consensus from the meeting this morning seems to be that we should fail the program if the probes can't be registered.